File: App/Resources/AlgorithmDetails/minheapsort/minheapsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int arr[], int n) {
  for (int i = 0; i < n; i++) {
    if (i == 0) {
      printf("[%d, ", arr[i]);
    } else if (i != n - 1) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d]", arr[i]);
    }
  }
}

void siftDown(int arr[], int root, int size) {
  while (1) {
    int smallest = root;
    int left = 2 * root + 1;
    int right = 2 * root + 2;
    if (left < size && arr[left] < arr[smallest]) {
      smallest = left;
    }
    if (right < size && arr[right] < arr[smallest]) {
      smallest = right;
    }
    if (smallest == root) {
      break;
    }
    swap(&arr[root], &arr[smallest]);
    root = smallest;
  }
}

void heapify(int arr[], int n) {
  for (int i = n / 2 - 1; i >= 0; i--) {
    siftDown(arr, i, n);
  }
}

void reverse(int arr[], int n) {
  int low = 0, high = n - 1;
  while (low < high) {
    swap(&arr[low], &arr[high]);
    low++;
    high--;
  }
}

void sort(int arr[], int n) {
  heapify(arr, n);
  for (int end = n - 1; end > 0; end--) {
    swap(&arr[0], &arr[end]);
    siftDown(arr, 0, end);
  }
  reverse(arr, n);
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/minheapsort/minheapsort.go
package main

import (
	"fmt"
)

func siftDown(arr []int, root, size int) {
	for {
		smallest := root
		left := 2*root + 1
		right := 2*root + 2
		if left < size && arr[left] < arr[smallest] {
			smallest = left
		}
		if right < size && arr[right] < arr[smallest] {
			smallest = right
		}
		if smallest == root {
			break
		}
		arr[root], arr[smallest] = arr[smallest], arr[root]
		root = smallest
	}
}

func heapify(arr []int) {
	n := len(arr)
	for i := n/2 - 1; i >= 0; i-- {
		siftDown(arr, i, n)
	}
}

func reverse(arr []int) {
	low, high := 0, len(arr)-1
	for low < high {
		arr[low], arr[high] = arr[high], arr[low]
		low++
		high--
	}
}

func sort(arr []int) []int {
	heapify(arr)
	for end := len(arr) - 1; end > 0; end-- {
		arr[0], arr[end] = arr[end], arr[0]
		siftDown(arr, 0, end)
	}
	reverse(arr)
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}
